### PR TITLE
Add the WooPay Express button to the Pay for order page

### LINKED ISSUE #5901: Add the WooPay Express button to the Pay for order page
### Description
Add the WooPay Express button to the Pay for order page.

[Reference P2 RFC: WooPay Missing “Pay for order” flow](https://woopayp2.wordpress.com/2023/03/21/extensions-incompatibility-missing-pay-for-order-flow/)

### Acceptance criteria

* Append the WooPay Express button to the Pay for order page.
* Pass the pay for order params to WooPay.

### Designs
[Reference comment](https://woopayp2.wordpress.com/2023/03/10/pt-express-button-compatibility-phase-2/#comment-2696)

### PR DESCRIPTION
**Pull request by hsingyuc**
Fixes #5901 

**Note: This PR ~~still needs design and~~ won't be merged until [WooPay](2104-gh-Automattic/woopay) and [WC](https://github.com/woocommerce/woocommerce/pull/37588) PRs are merged.**

#### Changes proposed in this Pull Request
1. Add the WooPay Express button to the pay-for-order page
2. Pass the pay-for-order params `order_id`, `pay_for_order`, `key`, and `billing_email` to WooPay redirect URL.

#### Screenshot ( ~~Needs design~~ )

Before             |  After
:-------------------------:|:-------------------------:
Screen Shot 2023-03-29 at 5 15 03 PM  | Screen Shot 2023-04-05 at 5 19 10 PM

#### Testing instructions
Screen Shot 2023-04-05 at 5 21 07 PM

1. Test with [WC](https://github.com/woocommerce/woocommerce/pull/37588) PR
2. Install and activate a plugin with pay-for-order flow. ex WooCommerce-Bookings
1. Purchase the product with pay-for-order flow
1. Go to WooCommerce -> Orders
1. Find the order you just replaced and order_data metabox
1. Click on the `Customer payment page →` link next to the order status
1. Check if the WooPay Express button is append on the pay for order page shown in the above picture
1. Copy the `order_id` and the `key` from the URL
1. Click on the WooPay Express button
1. It should redirect to the WooPay checkout page
1. Confirm and compare the URL containing the params `&pay_for_order=true&order_id=YOUR_ORDER_ID&key=wc_order_YOUR_KEY&billing_email=YOUR_BILLING_EMAIL`

*

-------------------

- [X] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5903` or branch name `add/pay-for-order` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/pay-for-order) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: dcec94200f9551ecf47b5e098cced779f04a10d7
- Build time: 2023-09-13 01:22:16 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by pierorocca** [comment]
Hi @hsingyuc thanks for the PR review request! :) I'm adding a link to the rough sketch Elizabeth did as a quick favor as they were 100% allocated to BNPL designs. https://woopayp2.wordpress.com/2023/03/10/pt-express-button-compatibility-phase-2/#comment-2734

In that approach the WooPay button is just above the Credit / Debit radio button.

**Review by elizaan36 — APPROVED** [review]
LGTM 🎉

**Comment by elizaan36** [comment]
Ah, just saw @pierorocca's comment. I think the approach shared in the PR works and is consistent with standard checkout, but placing the WooPay -OR- directly above the payment methods would be ideal. This will help shoppers associate WooPay with existing payment methods a bit better.

**Comment by pierorocca** [comment]
> Ah, just saw @pierorocca's comment. I think the approach shared in the PR works and is consistent with standard checkout, but placing the WooPay -OR- directly above the payment methods would be ideal. This will help shoppers associate WooPay with existing payment methods a bit better.

Echo this comment that placing it above the payment methods separated by an - OR - would be ideal specifically for the Pay for Order flow. Unlike standard checkout, a user may be coming back after many days or months have passed. They'd need to review the order first and then be presented with the "pay for order" payment options.

File: includes/class-wc-payments.php
Comment (pierorocca) [inline includes/class-wc-payments.php:1]: Thanks for the button placement change Emma!

File: includes/class-wc-payments-express-checkout-button-display-handler.php
Comment (malithsen) [inline includes/class-wc-payments-express-checkout-button-display-handler.php]: `is_payment_request_enabled` is only true for GPay and ApplePay flow. If the merchant only has WooPay EPB enabled, the button will not show up on the pay for order page. 
We can remove this condition if both flows now support the pay for order flow.

File: client/checkout/woopay/express-button/express-checkout-iframe.js
Comment (malithsen) [inline client/checkout/woopay/express-button/express-checkout-iframe.js]: This only adds the parameters when the user is WooPay-authenticated and automatically redirected to WooPay.
If we end up showing the OTP model, upon a successful authentication we will receive a `redirect_to_woopay_skip_session_init`. I think we should add the params when that happens as well.

**Review by malithsen — CHANGES_REQUESTED** [review]
Nice work @hsingyuc Thanks for working on this.

I haven't fully reviewed this or tested the flow end to end but wanted to leave a couple of preliminary observations that we may want to address before merging.

File: includes/class-wc-payments-express-checkout-button-display-handler.php
Comment (hsingyuc) [reply includes/class-wc-payments-express-checkout-button-display-handler.php]: >both flows now support the pay for order flow.

Do you mean `$is_payment_request_enabled` is for ApplePay and Gpay?  If so, I think we should change this condition to `$is_woopay_enabled`?

File: client/checkout/woopay/express-button/express-checkout-iframe.js
Comment (hsingyuc) [reply client/checkout/woopay/express-button/express-checkout-iframe.js]: Thank you for catching that! [122927c](https://github.com/Automattic/woocommerce-payments/pull/5903/commits/122927c9aeb4e046967f95175047fe9fb43b6522)

File: includes/class-wc-payments-express-checkout-button-display-handler.php
Comment (hsingyuc) [reply includes/class-wc-payments-express-checkout-button-display-handler.php]: I see what you mean! [5cc1e39](https://github.com/Automattic/woocommerce-payments/pull/5903/commits/5cc1e3988a33d3c4888ace22114ffa7ad2262c34)

**Review by malithsen — APPROVED** [review]
Looks good to me @hsingyuc Thanks for working on this!

**Comment by hsingyuc** [comment]
@malithsen I added [version control](https://github.com/Automattic/woocommerce-payments/pull/5903/commits/f8b4c17b63478db80245c6e7dc180421187d3de0) to make sure we only render the button when there are order and checkout order endpoints.  Since you are on the rotation and we need this PR to be merged to be able to test [this PR ](https://github.com/Automattic/woocommerce-payments/pull/7112) thoroughly, I'll merge this PR first but let me now if you have any concern.  Thank you.

**Comment by malithsen** [comment]
@hsingyuc Thanks. That's a good change. No concerns from me.

**Pull request merged**
